// @responsive-image/core/dist/index.js
function findMatchingImage(images, width, type) {
  const matchingImageWidth = images.map((i) => i.width).reduce((prevValue, w) => {
    if (w >= width && prevValue >= width) {
      return w >= prevValue ? prevValue : w;
    } else {
      return w >= prevValue ? w : prevValue;
    }
  }, 0);
  return images.find(
    (image) => image.width === matchingImageWidth && image.format === type
  );
}

const configNamespaces = /* @__PURE__ */ new Map();
function getConfig(namespace) {
  return configNamespaces.get(namespace);
}
function setConfig(namespace, config) {
  configNamespaces.set(namespace, config);
}

const env = {
  get screenWidth() {
    return typeof screen !== "undefined" ? screen.width : getConfig("env")?.fallbackScreenWidth ?? env.deviceWidths.at(-1) ?? 320;
  },
  get physicalWidth() {
    return env.screenWidth * env.devicePixelRatio;
  },
  get devicePixelRatio() {
    return typeof window !== "undefined" ? window?.devicePixelRatio ?? 1 : 1;
  },
  get deviceWidths() {
    return getConfig("env")?.deviceWidths ?? [
      640,
      750,
      828,
      1080,
      1200,
      1920,
      2048,
      3840
    ];
  }
};
function getDestinationWidthBySize(size = 100) {
  const factor = size / 100;
  return env.physicalWidth * factor;
}

function assert(message, check) {
  if (!check) {
    throw new Error(`Assertion failed: ${message}`);
  }
}

function resolveImage(data, { width, size, format } = {}) {
  const imageType = format ?? (Array.isArray(data.imageTypes) ? data.imageTypes[0] : data.imageTypes);
  const url = data.imageUrlFor(
    width ?? getDestinationWidthBySize(size),
    imageType
  );
  return url;
}

function getValueOrCallback(valueOrCb) {
  if (typeof valueOrCb === "function") {
    return valueOrCb();
  } else {
    return valueOrCb;
  }
}

export { assert, env, findMatchingImage, getConfig, getDestinationWidthBySize, getValueOrCallback, resolveImage, setConfig };
